Create spiral_overlay.py

ebmonson opened the pull request:
A modified version of overlay_test-cmap_scales.py with a slider for changing the overlay rotation in real time, a command line option for passing the file name, and more rigorous exception handling.

ebmonson committed:
Create spiral_overlay.py

A modified version of overlay_test-cmap_scales.py with a slider for changing the overlay rotation in real tim, a command line option for passing the file name, and more rigorous exception handling.

ebmonson committed:
Update spiral_overlay.py

Adds slider for pitch angle and radio buttons for chirality, scale choice, and arm number.

ebmonson committed:
Clean up spiral_overlay.py

ebmonson committed:
Update spiral_overlay.py

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,256 @@
+    if str(colorscale_option)=='n':
+        ax.imshow(galaxy_image,cmap='gray',origin='lower')
+    elif str(colorscale_option) == 'y':
+        # seismic & Greys recommended for images faint in log scale.
```
Add option to use seismic & Greys colormaps.

ebmonson committed:
Delete overlay_test-cmap_scales.py

ebmonson committed:
Merge pull request #2 from AGES-UARK/master

Merging docs

ebmonson committed:
Merge pull request #3 from ebmonson/master

Merge docs to spiral-overlay branch

ebmonson committed:
Update spiral_overlay.py

Added options for Seismic and Greys colormaps. Added check to allow filenames with .fits extension. Changed labels throughout. Added warning about performance decrease for large images to docstring.

ebmonson committed:
Update user_guide.rst

Added information about using spiral_overlay.py to check results.

ebmonson commented:
Should be ready to demo, possibly ready to merge.

jberlanga commented:
@ebmonson - A few suggestions:

Easy stuff:
- Make the spirals thicker.  When the window is maximized, it's harder to see the lines.
- Change the upper limit on pitch angle to 45 degrees.  We rarely see anything above 40, and this allows for finer control of the pitch angle at realistic scales.  Users can always dig into the code and change this for themselves if they _really_ want to see what an 89 degree pitch angle looks like.
- Change the default pitch angle to 20 degrees (this is close to the average & median in the local universe--see [Davis et al. 2015](http://adsabs.harvard.edu/abs/2015ApJ...802L..13D))
- Indicate that counter-clockwise chirality is considered negative & clockwise is considered positive by adding something like `(-)` & `(+)` next to those buttons.
- Change `Reset` button to something like `Reset rotation & pitch angle values`.  Users might expect other settings to change as well otherwise.

Things I'd like to see:
- Can we fix the x- and y-ranges so that the image isn't moving around when changing number of arms/rotation/etc.?  Maybe by default just go some percentage larger than the length & width--say 15-20%?
- Ability to set all the spiral arms to one color.  Let the user choose between multi-colored/white/black/red so that no matter what color scale they set, they can choose one spiral arm color that shows up easily.  I don't want to get rid of the multi-colored option since it's helpful when adjusting the pitch angle & rotation, but a single color will look better in figures for publications/posters/etc.
- Put all color-related boxes on the left & the arm number/chirality boxes on the right.
- Ability to set the origin of the spiral (e.g., make the minimum radius larger).  As you may have noticed, having the arms start at the center is a bit distracting when you have bars/large bulges.

One last thing--do we need the numbers on the x- & y- axes?  If we keep them, we should label them as pixels.  Although I'd prefer to lose them--I don't think they add any useful information to the end user.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+                        FITS filename can be passed as a command line argument,
+                        but the program will prompt for it if it isn't.
+    
+    OUTPUT:             Interactive plot in python window.
```
Technically Matplotlib window.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+                            tan(phi) = b
+                            phi = pitch angle
+    
+                        Edit theta_max, title, origin of the plot (imshow)
```
This is probably not something users need to edit in the source code any longer since they're working interactively instead of from the command line.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+                        Colormaps schemes available can be seen at:
+                        http://matplotlib.org/examples/color/colormaps_reference.html
+    
+                        Recommended cmap options are included as comments.
```
Ditto.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+                        Recommended cmap options are included as comments.
+                        
+                        It is recommended that you use cropped images with this program. There
+                        is a marked decrease in performance as image size increases.
```
Specify that the program will run slower, unless you mean that other aspects may not work as well, either.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+                        using the sliders below the plot. The reset button initializes the
+                        slider to default values (0.0 for rotation, 25.0 for pitch).
+                        
+                        Scale can be swapped between logarithmic and linear modes with the radio
```
Update to include all color scales available.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+                        Scale can be swapped between logarithmic and linear modes with the radio
+                        buttons on the left side of the plot. Defaults to linear scale.
+                        
+                        Chirality (corresponding to the sign of the pitch angle: CCW = +, CW = -)
```
CW is positive.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+
+#-----------------------------Default/initial values-----------------------------
+DEFAULT_ARMS = 2    #Between 1 and 6, inclusive
+DEFAULT_PITCH = 23.0  #Between about -89.12 and 89.12 Anything bigger causes overflow issues.
```
Make sure this matches the number you specify in the  docstring.

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+    if sys.argv[1][-4:] == ".fit" or sys.argv[1][-5:] == ".fits":
+        gal_file = sys.argv[1]
+    else:
+        print "Usage: python spiral_overlay.py <filename>"
```
There's a more Pythonic way to output error messages.  

Once you start doing things like write unit tests, if you get failing behavior, those tests won't necessarily catch it if you just use print statements. 

See [here](https://github.com/astropy/astroplan/blob/master/astroplan/plots/sky.py#L127) for an example (although note that this example doesn't necessitate exiting the program).

jberlanga commented on 2dfft_utils/analysis/spiral_overlay.py:
```diff
@@ -0,0 +1,276 @@
+try:
+    #pull the pixel values and save them as a uint8 numpy array
+    galaxy_image = fits.getdata(gal_file)
+except IOError:
```
This is better!

jberlanga commented on docs/user_guide.rst:
```diff
@@ -732,12 +732,21 @@ Automated Method
+The overlay can be saved by clicking the ``Save`` icon in the lower toolbar.
 
-  You can save the image by clicking on the ``Save`` button (shaped like a floppy disk) in the window that pops up.
+Default overlay and colormap options can be edited by modifying the constants at the top of the source file.
```
Maybe leave this out to avoid distracting less advanced users.

ebmonson committed:
Update spiral_overlay.py

-Added option to set spirals all one color
-Added save button that saves only the central figure
-Updated docstring
-Removed tick marks and labels from central figure
-Relabeled some options
-Changed max pitch to 45° and default to 20°

ebmonson committed:
Update user_guide.rst

ebmonson committed:
Update spiral_overlay.py

-Sped up plotting
-Eliminated global variables
-Moved overlay parameters into classes
-Modified image normalization to better accommodate real galaxy images

ebmonson committed:
Update user_guide.rst

ebmonson committed:
Updated for more consistent style
